File: frontend/src/types/listing.ts
export const PROPERTY_TYPES = [
  { value: 'tent', label: 'Палатка' },
  { value: 'yurt', label: 'Юрта' },
  { value: 'cabin', label: 'Домик' },
  { value: 'dome', label: 'Купол' },
  { value: 'treehouse', label: 'Дом на дереве' },
  { value: 'other', label: 'Другое' },
] as const

export type PropertyType = (typeof PROPERTY_TYPES)[number]['value']

export interface Listing {
  id: number
  title: string
  description: string
  price_per_night: string
  city: string
  region: string
  address: string
  property_type: PropertyType
  max_guests: number
  bedrooms: number
  amenities: string[]
  images: string[]
  status: string
  created_at: string
  updated_at: string
}

export interface ListingCreatePayload {
  title: string
  description: string
  price_per_night: number
  city: string
  region: string
  address: string
  property_type: PropertyType
  max_guests: number
  bedrooms: number
  status: 'draft' | 'active'
}

File: backend/app/models/listing.py
import enum
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Enum, Integer, Numeric, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class PropertyType(str, enum.Enum):
    TENT = "tent"
    YURT = "yurt"
    CABIN = "cabin"
    DOME = "dome"
    TREEHOUSE = "treehouse"
    OTHER = "other"


class ListingStatus(str, enum.Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    INACTIVE = "inactive"


class Listing(Base):
    __tablename__ = "listings"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False, default="")
    price_per_night: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    city: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    region: Mapped[str] = mapped_column(String(100), nullable=False, default="")
    address: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    property_type: Mapped[PropertyType] = mapped_column(
        Enum(PropertyType, name="property_type_enum", native_enum=True),
        nullable=False,
        index=True,
    )
    max_guests: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    bedrooms: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    amenities: Mapped[list] = mapped_column(JSONB, nullable=False, default=list)
    latitude: Mapped[float | None] = mapped_column(Numeric(9, 6), nullable=True)
    longitude: Mapped[float | None] = mapped_column(Numeric(9, 6), nullable=True)
    status: Mapped[ListingStatus] = mapped_column(
        Enum(ListingStatus, name="listing_status_enum", native_enum=True),
        nullable=False,
        default=ListingStatus.DRAFT,
        index=True,
    )
    promotion: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=False,
        index=True,
    )
    promotion_old_price: Mapped[float | None] = mapped_column(Numeric(10, 2), nullable=True)
    promotion_new_price: Mapped[float | None] = mapped_column(Numeric(10, 2), nullable=True)
    promotion_conditions: Mapped[str] = mapped_column(Text, nullable=False, default="")
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    listing_images: Mapped[list["ListingImage"]] = relationship(
        "ListingImage",
        back_populates="listing",
        cascade="all, delete-orphan",
        order_by="ListingImage.sort_order",
    )
